### ISSUE TITLE
How about replacing the softmax loss like Face Recognition Community did?

### ISSUE THREAD

**Issue by Jumabek** [AlexeyAB/darknet#512]
Hi @AlexeyAB ,

I read in multiple places that Face Recognition community replacing traditional Softmax with Margin Softmax, Angular Margin, Cosine Margin and and getting astounding improvement in accuracy. Latest paper about one of such methods: https://arxiv.org/pdf/1801.07698.pdf

From what I grasped, those methods work on reducing the intra class distance and maximizing the inter class distance. 

Would you be interested in implementing such loss functions for YOLO as well?

I myself trying as well, however, my progress is quite slow .

**Comment by AlexeyAB** [comment]
Hi, @Jumabek 

Did you try to implement additive angular margin (ArcFace) or CosineFace instead of Softmax?

|ArcFace|![image](https://user-images.githubusercontent.com/4096485/38168160-a655114a-354d-11e8-9f5a-e3a051ce2726.png)|
|---|---|
|Softmax|![image](https://user-images.githubusercontent.com/4096485/38168174-c7e7cb9a-354d-11e8-854b-c9060ce96f45.png)|




* Did you try to use CUDA-code of CosineFace(amsoftmax)? 
 https://github.com/deepinsight/insightface/blob/master/3rdparty/operator/amsoftmax.cu#L13-L36

* Or did you try to use Python code of ArcFace? https://github.com/deepinsight/insightface/blob/40cb9b223a487b6b3aa19f0316d9f7a0b1e28af4/src/train_softmax.py#L195-L234

----

As I understand the main problem in Softmax that they want to solve:

> However, the Softmax loss function does not explicitly optimise the features to have higher similarity score for positive pairs and lower similarity score for negative pairs, which leads to a performance gap.

So can ArcFace activation improve accuracy in detection tasks, or only in binary classification case?

![image](https://user-images.githubusercontent.com/4096485/38168149-4da95808-354d-11e8-88d9-bb008b715693.png)

**Comment by Jumabek** [comment]
@AlexeyAB , I am planning to implement CosineFace because it has similar performance with Arcface but mathematically a lot simpler. I specifically wanted to implement them in darkent because I have fire/smoke detector which simply does patch classification. After you question I now think maybe I am the only one who need it. Because it is for classification not for detection. 

I didnt know there is a cuda code.
I guess I will try to use CUDA-code of CosineFace(amsoftmax). In case of improvement in Fire/Smoke patch classification, I will report it here.
But that is not a promise, because as of now I switched to Face Recognition project.

Thank you for your attention.

**Comment by AlexeyAB** [comment]
@Jumabek For the interest, what framework do you use for face recognition? And what is the most advanced framework for face recognition now?

**Comment by Jumabek** [comment]
@AlexeyAB , 
According to [MegaFace](http://megaface.cs.washington.edu/results/facescrub.html) (challenging FR dataset)  there are 2 best methods in my fringe right now:
1. CosineFace - Caffe: https://github.com/happynear/AMSoftmax
2. ArcFace - MXNet: https://github.com/deepinsight/insightface

Currently I resolved to use CosineFace method of https://github.com/Joker316701882/Additive-Margin-Softmax implementation which is in python/TensorFlow
